### ISSUE TITLE
Normalise popularity scores by channel

### ISSUE THREAD

**Issue by LorenzoMinto** [brave/news-aggregator#611]

**Comment by aurangzaib048** [comment]
Do we need any changes in news-aggregator or its just for the reference? cc: @LorenzoMinto

**Comment by LorenzoMinto** [comment]
Yes, we should normalise the scores on the server-side before adding them to the feed. This will also act as a screen on using the raw popularity scores from Search.

**Comment by aurangzaib048** [comment]
As per my understanding, we need to normalize the score from `pop_score` 0-300 right?

**Comment by fallaciousreasoning** [comment]
@LorenzoMinto should we normalize by Publisher instead? Sources can be in multiple channels, and I'm not sure how the normalization would work there.

For the scores, I was thinking it would work something like this Rana:

```
target_max  = 300

for each publisher
    find the max score for this publisher

    # project to the range min_score to max_score
    for each article from publisher
         score = (article.score / max_article_score) * target_max
```

@LorenzoMinto would we want to also add a min score?

**Comment by aurangzaib048** [comment]
Or maybe something like this?
```
score = ((article.score – min(publisher_article_score)) / (max(publisher_article_score) – min(publisher_article_score))) * target_max
```

**Comment by mattmcalister** [comment]
If I understand the impact of this correctly then I would suggest normalizing at the channel level. We want major Crypto stories to have a chance to compete with major Top News stories, depending on what the user is Following. We don't care if the leading stories from each are both from the BBC, for example.

**Comment by aurangzaib048** [comment]
I just wanted to highlight an issue with our `Brave Search Popularity API`. If a publisher posts a new article called `Breaking news`, the API cannot give us the `pop_score` because, at that time, the API doesn’t have the data for that article. How can we handle that thing?

**Comment by fallaciousreasoning** [comment]
Okay @LorenzoMinto, that definitely makes sense - we could also average the normalized score across all channels the article is in :smile: 

I just thought of another little hiccup - articles can be present in the feeds for multiple reasons, and different regions can have different channels for the same publisher. Would we do the same thing here, and just take the highest pop_score?

:+1: to having a score we cutoff at, some of the scores are absurd, and if we normalize using that it'll make all the other articles in that channel sink way down the feed. Alternatively, we could use `logn(1 + score)` to have increases in score worth less, the higher it gets?

**Comment by LorenzoMinto** [comment]
I think we should solve the multiple locales issue separately, as it's an independent source of confusion that maybe we could handle better upstream.

`logn(1+score)` it's also a good option, although we might not want to flatten scores too much in a certain range. So maybe it could be linear up to `score_max` and then switch into logarithmic

@aurangzaib048 for articles without popularity I would just treat them as they come initially, then we could figure out a way to predict popularity, I have something in mind for that.

**Comment by fallaciousreasoning** [comment]
Yeah that sounds perfect:
[0, 300]: Linear
[300, ]: 300 + logn(score) - any ideas for a good log base/multiplier?

**Comment by LorenzoMinto** [comment]
Actually instead of logarithm we could use a root function, to have it decay more softly. So we could do something like:
```
[0,300]: y = x
[300, ]: y = 299 + (1 + x - 300)^(2/3)
```

**Comment by aurangzaib048** [comment]
As per my understanding, we need `pop_score` of article by channels, in some range for example [0:300]. It will be achievable by the following steps:
-  Get the `pop_score` for all the articles
- Give the max score (for example 300) to the most recent _**(If you wanted less than an hour old)**_
- Then calculate rest of the article `pop_score` according to some formula.

P.S. Can you please explain the formula you mentioned above, what do you means by `x` and `y`? So that I can proceed accordingly. cc: @LorenzoMinto

**Comment by fallaciousreasoning** [comment]
`Y` is the normalized `pop_score` and `X` is the non-normalized `pop_score`, from Brave Search

Basically, for scores less than or equal to 300, we'll just use the score from Brave Search. After 300, we want to avoid massive scores (like 10000), so we have a decay function.

```
cutoff = 300
exponent = 2/3

if pop_score < cutoff:
    return pop_score
return (cutoff - 1) + (1 + pop_score - cutoff)^exponent
```

**Comment by LorenzoMinto** [comment]
Thanks @fallaciousreasoning. To complete the process, after we have the cutoff scores we can do min-max normalisation and project the pop_scores onto 0 to 100 (`target_score`).

**Comment by aurangzaib048** [comment]
Deployed in staging.

**Comment by aurangzaib048** [comment]
Deployed.
